docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,75 @@
+# Generics parameters of ByRefLike types
+
+Using ByRefLike types in Generic parameters is possible by building upon support added for `ref` fields. Scenarios that would benefit most from this are those involving `Span<T>`. For example, consider the following examples:
+
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
```

Comment:
Will this have a type loader impact too or it's more about just relaxing validation in the type loader and defining behaviors of instructions when T is a byref-like type?

It kind of feels like we could also handle this in the runtime the same way we handle the existing constraints (`struct`, `new()`, etc.). But maybe validating this can be purely a tooling concern, not runtime concern, since we will have semantics for when IL operations don't make sense for a specific T.

```csharp
class Foo<AnyT> where AnyT : any
{
    // Do we get a TypeLoadException for List<AnyT> when JIT compiling this method
    // or do we get an InvalidProgramException when the List<T> constructor is trying
    // to get an empty array of T for the backing store?
    static object Get() => new List<AnyT>();
}
```

Reply:
> Will this have a type loader impact too or it's more about just relaxing validation in the type loader and defining behaviors of instructions when T is a byref-like type?

The latter. My thought here was to push failure to JIT time. In the scenario above, if `Get()` was never called then it should be okay. Implementers could use the `IsByRefLike<T>()` to tailor the failure case if desired though.

Reply:
```
class Foo<AnyT> where AnyT : any
{
    // Do we get a TypeLoadException for List<AnyT> when JIT compiling this method
    // or do we get an InvalidProgramException when the List<T> constructor is trying
    // to get an empty array of T for the backing store?
    static object Get() => new List<AnyT>();
}
```
I expect that this will be C# compile error. (I do not expect we will add `any` relaxation for `List<T>`.)

If you agree, the exact runtime behavior is not really interesting here. I can also just crash somewhere in the runtime like it is the case for invalid IL in general.

Reply:
> I expect that this will be C# compile error

I expect that too - this was pseudo-C# just to convey what I mean :).

> If you agree, the exact runtime behavior is not really interesting here.

The runtime behavior might be observable through `MakeGenericType`/`MakeGenericMethod` - if we allow `MakeGeneric` with byref-like types and someone uses it with a type where the `T` is not `any` and we don't block it, they'll get an oddly damaged type. If we do block it, the question is where to block it. The place where we validate constraints seems like a natural fit. But it will affect the above mentioned code paths too.